### PR TITLE
its-tracking: put rest of tracking inside traits

### PR DESCRIPTION
**Pull request by mconcas**
This PR is in preparation for #9289, to be preferably merged earlier than that. 
This change is in view of adding corresponding GPU versions of rest of tracker functions.

- All steps of ITS tracking are now steered by `TrackerTraits` class.
- `Tracker` class is now just an interface to steer traits and manage MC associations via `TimeFrame`.
- Removed unused `MemoryParams` class
- Cleaned up some unused files such as PrimaryVertexContext.{cxx,h}
 
Pinging @mpuccio and @shahor02 for review.

### CODE REVIEW

**Comment by davidrohr** [comment]
@mconcas : sorry: I was constantly delaying to merge this, but it causes merge conflicts ....
Could you rebase. @mpuccio : could you approve, then we'll merge once it is green.

**Comment by mconcas** [comment]
> @mconcas : sorry: I was constantly delaying to merge this, but it causes merge conflicts .... Could you rebase. 

@davidrohr : No prob. Currently rebasing GPU one (#9289). This is the next. then the merging order is at your convenience. Some x-rebasing would be required anyways.

**Comment by mconcas** [comment]
By the way, I'd also wait for #9492 to be merged

**Comment by mconcas** [comment]
Hi @mpuccio @shahor02 @davidrohr:

Tested locally on single TF 200 pp events:

**mconcas:fully-separate-traits (this PR):**

```bash
[2846666:its-tracker]: [15:34:30][INFO] Loading ITS L2G matrices from TGeo
[2846666:its-tracker]: [15:34:30][INFO]  - rejected 0/63 ROFs: random/mult.sel:0, vtx.sel:0
[2846666:its-tracker]: [15:34:30][INFO]  - Vertex seeding total elapsed time: 60.13393 ms for 180 vertices found in 63 ROFs
[2846666:its-tracker]: [15:34:30][INFO]  - Beam position computed for the TF: 0.0015631894, -0.00079381414
[2846666:its-tracker]: [15:34:30][INFO]  - Timeframe initialisation completed in: 61.2386 ms
[2846666:its-tracker]: [15:34:30][INFO]  - Tracklet finding completed in: 27.1146 ms
[2846666:its-tracker]: [15:34:30][INFO]  - Number of tracklets: 27756
[2846666:its-tracker]: [15:34:30][INFO]  - Cell finding completed in: 0.518735 ms
[2846666:its-tracker]: [15:34:30][INFO]  - Number of Cells: 17515
[2846666:its-tracker]: [15:34:30][INFO]  - Neighbour finding completed in: 0.480116 ms
[2846666:its-tracker]: [15:34:30][INFO]  - Road finding completed in: 0.175459 ms
[2846666:its-tracker]: [15:34:30][INFO]  - Number of Roads: 3413
[2846666:its-tracker]: [15:34:31][INFO]  - Track finding completed in: 322.609 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Extending tracks completed in: 0.00032 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Timeframe initialisation completed in: 0.166479 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Tracklet finding completed in: 24.4006 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Number of tracklets: 18901
[2846666:its-tracker]: [15:34:31][INFO]  - Cell finding completed in: 0.275928 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Number of Cells: 8051
[2846666:its-tracker]: [15:34:31][INFO]  - Neighbour finding completed in: 0.187968 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Road finding completed in: 0.03257 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Number of Roads: 519
[2846666:its-tracker]: [15:34:31][INFO]  - Track finding completed in: 48.6717 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Extending tracks completed in: 0.00017 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Timeframe initialisation completed in: 0.086189 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Tracklet finding completed in: 32.6333 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Number of tracklets: 39439
[2846666:its-tracker]: [15:34:31][INFO]  - Cell finding completed in: 0.723744 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Number of Cells: 15274
[2846666:its-tracker]: [15:34:31][INFO]  - Neighbour finding completed in: 0.400477 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Road finding completed in: 0.441327 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Number of Roads: 9348
[2846666:its-tracker]: [15:34:31][INFO]  - Track finding completed in: 511.156 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Extending tracks completed in: 0.00037 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Short primaries finding completed in: 7e-05 ms
[2846666:its-tracker]: [15:34:31][INFO]  - Timeframe 0 processing completed in: 1031.31ms using 1 threads.
[2846666:its-tracker]: [15:34:31][INFO] ITSTracker pushed 3602 tracks and 180 vertices
[2846666:its-tracker]: [15:34:31][INFO] ITSTracker pushed 3602 track labels
```
**AliceO2Group:dev:**

```bash
[2871750:its-tracker]: [15:57:12][INFO] Loading ITS L2G matrices from TGeo
[2871750:its-tracker]: [15:57:13][INFO]  - rejected 0/63 ROFs: random/mult.sel:0, vtx.sel:0
[2871750:its-tracker]: [15:57:13][INFO]  - Vertex seeding total elapsed time: 60.861504 ms for 180 vertices found in 63 ROFs
[2871750:its-tracker]: [15:57:13][INFO]  - Beam position computed for the TF: 0.0015631894, -0.00079381414
[2871750:its-tracker]: [15:57:13][INFO]  - Timeframe initialisation completed in: 61.9965 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Tracklet finding completed in: 26.6996 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of tracklets: 27756
[2871750:its-tracker]: [15:57:13][INFO]  - Cell finding completed in: 0.518756 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of Cells: 17515
[2871750:its-tracker]: [15:57:13][INFO]  - Neighbour finding completed in: 0.536165 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Road finding completed in: 0.172928 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of Roads: 3413
[2871750:its-tracker]: [15:57:13][INFO]  - Track finding completed in: 326.035 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Extending tracks completed in: 6.9e-05 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Timeframe initialisation completed in: 0.185389 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Tracklet finding completed in: 23.9513 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of tracklets: 18901
[2871750:its-tracker]: [15:57:13][INFO]  - Cell finding completed in: 0.284638 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of Cells: 8051
[2871750:its-tracker]: [15:57:13][INFO]  - Neighbour finding completed in: 0.229058 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Road finding completed in: 0.0308 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of Roads: 519
[2871750:its-tracker]: [15:57:13][INFO]  - Track finding completed in: 47.8127 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Extending tracks completed in: 6e-05 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Timeframe initialisation completed in: 0.09657 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Tracklet finding completed in: 31.7958 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of tracklets: 39439
[2871750:its-tracker]: [15:57:13][INFO]  - Cell finding completed in: 0.740454 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of Cells: 15274
[2871750:its-tracker]: [15:57:13][INFO]  - Neighbour finding completed in: 0.385507 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Road finding completed in: 0.407717 ms
[2871750:its-tracker]: [15:57:13][INFO]  - Number of Roads: 9348
[2871750:its-tracker]: [15:57:14][INFO]  - Track finding completed in: 515.067 ms
[2871750:its-tracker]: [15:57:14][INFO]  - Extending tracks completed in: 0.0002 ms
[2871750:its-tracker]: [15:57:14][INFO]  - Short primaries finding completed in: 5e-05 ms
[2871750:its-tracker]: [15:57:14][INFO]  - Timeframe 0 processing completed in: 1036.95ms using 1 threads.
[2871750:its-tracker]: [15:57:14][INFO] ITSTracker pushed 3602 tracks and 180 vertices
[2871750:its-tracker]: [15:57:14][INFO] ITSTracker pushed 3602 track labels
```

At your review.

File: Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h
Comment (shahor02) [inline Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h]: this is just an enum, no point in passing by reference.

File: Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h
Comment (shahor02) [inline Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h:108]: idem

**Review by shahor02 — DISMISSED** [review]
for me looks fine, just minor cosmetic comments.

File: Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h
Comment (mconcas) [reply Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h]: Sure, changed.

File: Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h
Comment (mconcas) [reply Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h:108]: Not as trivial as it seems, somehow this raises errors in template variadic parameters expansion for the wrapper `evaluateTask(...)`.

E.g:
```bash
home/mconcas/alice/sw/SOURCES/O2/fully-separate-traits/0/Detectors/ITSMFT/ITS/tracking/include/ITStracking/Tracker.h:91:9: note:   template argument deduction/substitution failed:
2022-07-23@16:45:59:DEBUG:O2:O2:0: /home/mconcas/alice/sw/SOURCES/O2/fully-separate-traits/0/Detectors/ITSMFT/ITS/tracking/src/Tracker.cxx:53:103: note:   inconsistent parameter pack deduction with 'int' and 'int&'
```

Hence I decided not to touch too much.

**Review by mpuccio — APPROVED** [review]
@davidrohr this can be merged as soon as you want. It works on both CTF from P2 and MC giving the same results

**Pull request merged**
